src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Handler.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClient.SNI;
+using Microsoft.Data.SqlClientX.IO;
+
+
+namespace Microsoft.Data.SqlClientX.Handlers
+{
+    /// <summary>
+    /// Interface to repreent the handler.
+    /// </summary>
+    /// <typeparam name="TRequest"></typeparam>
+    internal interface IHandler<TRequest> where TRequest : HandlerRequest
+    {
+        /// <summary>
+        /// The call to handler to execute the request.
+        /// </summary>
+        /// <param name="request"></param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        public abstract ValueTask Handle(TRequest request, bool isAsync, CancellationToken ct);
+
+        /// <summary>
+        /// If a next handler is to be executed in the chain, then SetNext is used for those purposes.
+        /// </summary>
+        /// <param name="handler"></param>
+        public void SetNext(IHandler<TRequest> handler);
+    }
+
+
+    /// <summary>
+    /// The abstract class representing the handler.
+    /// 
+    /// </summary>
+    /// <typeparam name="TRequest"></typeparam>
+    internal abstract class Handler<TRequest> : IHandler<TRequest> where TRequest : HandlerRequest
+    {
+        /// <summary>
+        /// Stores the internal handler to be executed next.
+        /// </summary>
+        private IHandler<TRequest> _nextHandler;
+
+        /// <summary>
+        /// Property for interacting with the Next Handler.
+        /// </summary>
+        public IHandler<TRequest> NextHandler { get => _nextHandler; set => _nextHandler = value; }
```

Comment:
Shouldn't this just be an auto-property?

Reply:
Actually I think this should be virtual. The base can provide an auto-property and the subclasses can do more complex stuff .